File: packages/server/javascript/src/context.ts
/**
 * ABTO request context propagation for Node.js.
 *
 * Server SDK scope: thin helper only. It routes provider SDK calls through the
 * ABTO Gateway and carries the gateway identifier headers. The ABTO transport owns
 * only direct fallback for safely classified Gateway availability failures. It
 * does not classify or retry OpenAI or model-provider errors; the official OpenAI
 * SDK remains the retry authority. The Gateway remains the source of truth for
 * token, cost, latency, request_id, and variant assignment.
 *
 * Gateway identifier contract:
 *   - device_id -> header `x-abto-device-id` (optional; without it the call still
 *                 succeeds but drops out of user-level analytics and sticky assignment)
 *   - node_key -> header `x-abto-node-key`  (required by the gateway). "feature.node"
 *                 dot notation, e.g. "resume.make"; feature is the read-time prefix.
 *   - trace_id -> W3C `traceparent`. End-user action bundle; gateway-deferred (Round1).
 *
 * request_id (PK, returned as `x-abto-request-id`), tenant_id (from API key), and
 * variant_id (A/B assignment) are gateway-owned and are NOT sent by the SDK.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import { newUuidV7TraceId, randomHex } from './uuid.js';

/** Gateway-facing header names. Only what the gateway reads. */
export const ABTO_HEADER = {
  deviceId: "x-abto-device-id",
  nodeKey: "x-abto-node-key",
  traceparent: "traceparent",
} as const;

export interface AbtoContext {
  /** Device id. Sent as `x-abto-device-id` (optional for the gateway). */
  deviceId?: string;
  /** "feature.node" dot-notation node key (e.g. "resume.make"). Sent as `x-abto-node-key`. */
  nodeKey?: string;
  /** End-user action bundle id. Sent as W3C `traceparent`. Gateway-deferred. */
  traceId?: string;
}

export interface BuildHeaderOptions {
  /** Emit a W3C traceparent derived from traceId. Defaults to true when traceId is set. */
  includeTraceparent?: boolean;
}

const storage = new AsyncLocalStorage<AbtoContext>();

/** 32-hex-char trace id, per W3C trace-context. */
export function createTraceId(): string {
  return newUuidV7TraceId();
}

/** Build a W3C `traceparent` value for the given trace id. */
export function createTraceparent(traceId: string): string {
  return `00-${traceId}-${randomHex(8)}-01`;
}

export function runWithAbtoContext<T>(ctx: AbtoContext, fn: () => T): T {
  const parent = storage.getStore();
  const merged: AbtoContext = { ...parent, ...ctx };
  return storage.run(merged, fn);
}

export function getAbtoContext(): AbtoContext | undefined {
  return storage.getStore();
}

export function getAbtoHeaders(
  ctx?: AbtoContext,
  options: BuildHeaderOptions = {},
): Record<string, string> {
  const c = ctx ?? storage.getStore() ?? {};
  const headers: Record<string, string> = {};
  if (c.deviceId) headers[ABTO_HEADER.deviceId] = c.deviceId;
  if (c.nodeKey) headers[ABTO_HEADER.nodeKey] = c.nodeKey;
  if (c.traceId && (options.includeTraceparent ?? true)) {
    headers[ABTO_HEADER.traceparent] = createTraceparent(c.traceId);
  }
  return headers;
}

File: packages/server/python/abto/context.py
"""Request-scoped ABTO identifier context for the `abto` Python package.

Mirrors @abto-app/calling: carries the gateway identifier headers via
contextvars so outbound provider calls can attach them. The ABTO transport owns
only direct fallback for safely classified Gateway availability failures. It
does not classify or retry OpenAI or model-provider errors; the official OpenAI
SDK remains the retry authority. The gateway remains the source of truth for
token, cost, latency, request_id, and variant assignment.
"""

from __future__ import annotations

import contextvars
import secrets
from contextlib import contextmanager
from dataclasses import dataclass, replace
from typing import Any, Dict, Iterator, Optional

ABTO_HEADER = {
    "device_id": "x-abto-device-id",
    "feature_id": "x-abto-feature-id",
    "traceparent": "traceparent",
}


@dataclass(frozen=True)
class AbtoContext:
    """Device id, dot-separated feature id, and end-user action trace id."""

    device_id: Optional[str] = None
    feature_id: Optional[str] = None
    trace_id: Optional[str] = None


_current: contextvars.ContextVar[AbtoContext] = contextvars.ContextVar(
    "abto_context", default=AbtoContext()
)


def get_context() -> AbtoContext:
    return _current.get()


_UNSET = object()


@contextmanager
def with_context(
    device_id: Any = _UNSET,
    feature_id: Any = _UNSET,
    trace_id: Any = _UNSET,
) -> Iterator[AbtoContext]:
    patch = {
        k: v
        for k, v in dict(
            device_id=device_id, feature_id=feature_id, trace_id=trace_id
        ).items()
        if v is not _UNSET
    }
    merged = replace(_current.get(), **patch)
    token = _current.set(merged)
    try:
        yield merged
    finally:
        _current.reset(token)


def create_trace_id() -> str:
    """32-hex-char trace id, per W3C trace-context."""
    return secrets.token_hex(16)


def create_traceparent(trace_id: str) -> str:
    return f"00-{trace_id}-{secrets.token_hex(8)}-01"


def get_headers(ctx: Optional[AbtoContext] = None) -> Dict[str, str]:
    c = ctx if ctx is not None else _current.get()
    headers: Dict[str, str] = {}
    if c.device_id:
        headers[ABTO_HEADER["device_id"]] = c.device_id
    if c.feature_id:
        headers[ABTO_HEADER["feature_id"]] = c.feature_id
    if c.trace_id:
        headers[ABTO_HEADER["traceparent"]] = create_traceparent(c.trace_id)
    return headers
